janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphVertexStep.java

```diff
@@ -85,20 +89,68 @@ private void initialize() {
             if (!starts.hasNext()) {
                 throw FastNoSuchElementException.instance();
             }
-            final JanusGraphMultiVertexQuery multiQuery = JanusGraphTraversalUtil.getTx(traversal).multiQuery();
             final List<Traverser.Admin<Vertex>> vertices = new ArrayList<>();
             starts.forEachRemaining(v -> {
                 vertices.add(v);
-                multiQuery.addVertex(v.get());
             });
             starts.add(vertices.iterator());
-            assert vertices.size() > 0;
-            makeQuery(multiQuery);
+            initializeMultiQuery(vertices);
+        }
+    }
 
-            multiQueryResults = (Vertex.class.isAssignableFrom(getReturnClass())) ? multiQuery.vertices() : multiQuery.edges();
+    /**
+     * This initialisation method is called the first time this instance is used and also when
+     * an attempt to retrieve a vertex from the cached multiQuery results doesn't find an entry.
+     * @param vertices A list of vertices with which to initialise the multiQuery
+     */
+    private void initializeMultiQuery(final List<Traverser.Admin<Vertex>> vertices) {
+        assert vertices.size() > 0;
+        List<Admin<Vertex>> parentStarts = new ArrayList<>();
+        if (vertices.size() == 1) {
+            if (multiQueryResults != null && multiQueryResults.containsKey(vertices.get(0).get())) {
+                return; // Nothing to do this vertex is already in the multiQueryResults
+            } else {
+                parentStarts = getParentMultiQueryStepStarts();
+            }
+        }
+        final JanusGraphMultiVertexQuery multiQuery = JanusGraphTraversalUtil.getTx(traversal).multiQuery();
+        vertices.forEach(v -> multiQuery.addVertex(v.get()));
+        parentStarts.forEach(v -> multiQuery.addVertex(v.get()));
+        makeQuery(multiQuery);
+
+        Map<JanusGraphVertex, Iterable<? extends JanusGraphElement>> results = (Vertex.class.isAssignableFrom(getReturnClass())) ? multiQuery.vertices() : multiQuery.edges();
+        if (multiQueryResults == null) {
+            multiQueryResults = results;
+        } else {
+            multiQueryResults.putAll(results);
+            results.clear();
```

Comment:
Same here. Just would like to know why do we call `clear` here.

Reply:
Good Question, and given that the results goes out of scope shortly afterwards we don't need it, I've removed it.